[#15] 장바구니

seohoon-dev opened the pull request:
- 장바구니에 아이템 추가 및 삭제
- 장바구니 목록 불러오기

seohoon-dev committed:
[#15] 장바구니

- 장바구니에 아이템 추가 및 삭제
- 장바구니 목록 불러오기

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/CartService.java:
```diff
@@ -0,0 +1,39 @@
+        Object value = listOperations.index(loginUserEmail, id);
+
+        if (value == null) {
+            throw new IllegalStateException("아이템 정보가 없습니다.");
```
이런 경우라면 사용자가 잘못된 id를 입력했을 가능성이 높으니 `IllegalArgumentException`이 맞지 않을까요?

f-lab-dev commented on src/main/java/com/hoon/foodrocket/config/RedisConfig.java:
```diff
@@ -0,0 +1,37 @@
+        RedisTemplate<String, Object> redisTemplate = new RedisTemplate<>();
+        redisTemplate.setConnectionFactory(redisConnectionFactory());
+        redisTemplate.setKeySerializer(new StringRedisSerializer());
+        redisTemplate.setValueSerializer(new GenericJackson2JsonRedisSerializer(new ObjectMapper()));
```
이 각각의 `serializer`들이 어떤 역할을 하는지 알아보시고 주석으로 설명을 달아주시면 좋을 것 같습니다.
`ObjectMapper`의 역할도 달아주시면 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/config/RedisConfig.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Bean
+    public RedisConnectionFactory redisConnectionFactory() {
+        return new LettuceConnectionFactory(redisProperties.getHost(),redisProperties.getPort());
```
여기서는 Lettuce를 쓰고있습니다. Lettuce 외에도 jedis, redisson 등 다른 클라이언트 라이브러리들이 많은데요. 각각 장단점을 비교해보시면 좋을 것 같습니다

f-lab-dev replied:
아 그리고 그 비교한거를 블로그에 정리해보시면 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/CartController.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/cart")
```
항상 복수형으로 해주시는게 좋습니다. `/carts` 처럼요

f-lab-dev reviewed (commented):
전체 소스를 다시 훑어보니까 레디스 세션을 적용한 곳이 안보이네요.
혹시 빠져있는거면 이것도 적용해주시면 좋을 것 같아요~

seohoon-dev committed:
[#14] 주문

- 주문 등록
- 주문 시 결제 진행
- 주문 완료 후 장바구니 비움

seohoon-dev committed:
리뷰 내용 반영

- 예외 유형 변경
- serializer와 objectmapper의 역할 주석으로 추가
- url 복수형으로 표기

f-lab-dev commented on src/main/java/com/hoon/foodrocket/config/RedisConfig.java:
```diff
@@ -22,9 +20,24 @@
     }
 
+    /**
+     * Redis는 직렬화된 데이터를 메모리 서버에 저장하는 형태를 가진 시스템으로 key, value는 직렬화가 필요하다.
```
직렬화가 꼭 필요할까요? 직렬화를 하는 이유가 뭘까요?
레디스는 데이터를 어떤 형식으로 저장하는지, 직렬화의 역할이 무엇인지 알아보시면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/config/RedisConfig.java:
```diff
@@ -22,9 +20,24 @@
+     * StringRedisSerializer는 단순 문자열을 바이트 배열로 변환할 수 있고, 기본 charset이 UTF-8이다.
+     * key값의 경우 주로 문자열을 사용하기 때문에 StringRedisSerializer를 사용했다.
+     * <p>
+     * GenericJackson2JsonRedisSerializer은 동적 타입을 사용하여 객체를 json에 매핑한다.
```
`동적 타입`과 `json에 매핑`이 더 자세하게 무엇을 의미하는걸까요? 이런 문장은 더 쉽게 풀어주는것이 좋습니다~

seohoon-dev committed:
리뷰 내용 반영

- 패키지명 변경(application -> service)
- 주문, 결제 순서 변경
- 결제 파트에 전략패턴 적용
- 주문 내역 시 커서 기반 페이징 적용
- 주문 상세 내역 추가

seohoon-dev committed:
Merge branch 'issue/15' of https://github.com/f-lab-edu/food-rocket into issue/14

seohoon-dev committed:
리뷰 내용 반영

- single import 방식으로 변경
- url 변경

seohoon-dev committed:
리뷰 내용 반영

- 쿼리에 주석 작성

seohoon-dev committed:
Merge pull request #18 from f-lab-edu/issue/14

[#14] 주문

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OrderController.java:
```diff
@@ -20,21 +22,62 @@
+        String loginUserEmail = HttpSessionUtil.getLoginUserEmail(session);
+
+        cardOrder.setUserEmail(loginUserEmail);
+        orderService.registerOrder(cardOrder, cardOrder.getCardPayment(), loginUserEmail, new CardPaymentProcess());
```
`new CardPaymentProcess()` 항상 호출마다 객체를 생성할 필요는 없습니다.
`CardPaymentProcess`는 **stateless**하므로 매번 객체를 새로 생성할 필요가 없습니다.
`stateless`에 대해 검색해보시면 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/service/OrderService.java:
```diff
@@ -0,0 +1,70 @@
+     * 2. 결제 진행
+     * 주문을 위해 전달받은 오더 파라미터를 paymentProcess()에 넘긴다.
+     * process()는 오더에 정의된 결제방식에 따라 결제를 진행한다.
+     * 결제 완료 후 결과값을 반환한다.
```
결제 실패 시나리오도 보완해주시면 좋을 것 같아요

seohoon-dev replied:
예외처리에 대한 걸 말씀하시는건가요?

f-lab-dev commented on src/main/java/com/hoon/foodrocket/service/OrderService.java:
```diff
@@ -0,0 +1,70 @@
+        // 결제 진행
+        boolean result = paymentProcessInterface.paymentProcess(payment);
+
+        if (result) {
```
이쪽쯤에서 `updateOrderStatus `를 호출해야할 것 같은데 `updateOrderStatus `는 언제 호출되나요?

seohoon-dev replied:
주문 등록은 mysql에서 기본값으로 'ORDER_REQUEST'을 주고 있습니다.
이후에 OrderController에서 주문에 대한 상태를 변경하도록 하고 있습니다.

f-lab-dev commented on src/main/resources/mapper/OrderMapper.xml:
```diff
@@ -3,33 +3,71 @@
+<!--    ex) 후라이드 치킨 외 2개-->
+<!--    각 주문에는 대응하는 주문 메뉴가 있기 때문에 INNER JOIN을 사용했고, 그룹화를 통해 주문별 첫번째 메뉴와 메뉴 갯수를 가져오도록 했습니다.-->
+<!--    화면에 보여지는 데이터는 한정적이기 때문에 커서 기반 페이징을 통해 불필요하게 많은 데이터를 가져오지 않도록 했으며-->
+<!--    최신 데이터가 먼저 보여지도록 날짜별로 정렬했습니다.-->
```
설명이 깔끔하군요👍

seohoon-dev committed:
[#19] paymentprocess 스프링 빈 등록

seohoon-dev committed:
[#21] 주문 상태 추가(결제_완료) 및 적용

seohoon-dev committed:
Merge pull request #23 from f-lab-edu/issue/21

[#21] 주문 상태 추가(결제_완료) 및 적용

seohoon-dev committed:
Merge pull request #22 from f-lab-edu/issue/19

[#19] paymentprocess 스프링 빈 등록

f-lab-dev commented on src/main/java/com/hoon/foodrocket/service/payment/CardPaymentProcess.java:
```diff
@@ -1,7 +1,14 @@
 
+/**
+ * 기존에 OrderController 에서 메소드를 호출할때마다 결제 프로세스 객체를 새로 생성해서 넘겨줬는데
+ * 결제 프로세스는 별도의 상태를 저장하지 않으므로 객체를 새로 생성하는 것이 서버 입장에서 부담이기 때문에
```
왜 부담인지 상세한 이유를 적어주시면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/service/payment/CardPaymentProcess.java:
```diff
@@ -1,7 +1,14 @@
+/**
+ * 기존에 OrderController 에서 메소드를 호출할때마다 결제 프로세스 객체를 새로 생성해서 넘겨줬는데
+ * 결제 프로세스는 별도의 상태를 저장하지 않으므로 객체를 새로 생성하는 것이 서버 입장에서 부담이기 때문에
+ * 이를 방지하고자 결제 프로세스를 스프링 빈에 등록시킴으로써 싱글톤으로 관리하도록 함
```
스프링 빈에 등록하는 이유는 싱글톤을 만들기 위함만은 아닙니다. 대부분의 클래스는 빈으로 등록하는데요. 다른 이유도 같이 들어주시면 좋을 것 같네요~

seohoon-dev committed:
[#19] paymentprocess 스프링 빈 등록

결제 프로세스를 스프링 빈으로 등록한 이유를 주석으로 작성

seohoon-dev committed:
[#20] UserService 테스트 작성

seohoon-dev committed:
리뷰 내용 반영

- 서버에 부담이 발생하는 이유 작성

seohoon-dev committed:
리뷰 내용 반영

- 추상적인 변수명 수정
- 도메인 클래스 안에 암호화 메서드 추가

seohoon-dev committed:
[#26] OrderService 테스트 작성

seohoon-dev committed:
Merge pull request #24 from f-lab-edu/issue/19

[#19] paymentprocess 스프링 빈 등록

seohoon-dev committed:
Merge pull request #27 from f-lab-edu/issue/26

[#26] OrderService 테스트 작성

seohoon-dev committed:
리뷰 내용 반영

seohoon-dev committed:
리뷰 내용 반영

- UserService 클래스에 암호화 메소드 구현

seohoon-dev committed:
Merge pull request #25 from f-lab-edu/issue/20

[#20] UserService 테스트 작성
